lib/vm-agent.js

```diff
@@ -9,450 +9,696 @@
  */
 
 /*
- * vm-agent.js
+ * Overview of Responsibilities
+ * ============================
+ *
+ * The job of vm-agent is to tell VMAPI about changes to VMs that occur on the
+ * CN. Except in case of errors, operations performed on the CN through SDC APIs
+ * should not need this, as cn-agent tasks do a load of the VM state after each
+ * action is performed and the results are put into VMAPI.
+ *
+ * The primary cases where vm-agent is solely responsible for letting SDC know
+ * about changes are:
+ *
+ *  * In the case of errors where a process crashes before updating VMAPI but
+ *    after making a change, or in the case of inconsistency due to other bugs.
+ *    The data on the CN is always treated as more correct and VMAPI should be
+ *    updated with the actual state on the CN.
+ *
+ *  * In case of changes performed manually by operators. When an operator makes
+ *    a change to a VM from the GZ on the CN, the APIs will not know about this
+ *    change unless told by vm-agent.
+ *
+ *  * When operations are triggered from within VMs themselves. This includes
+ *    shutdown and restart operations performed from within the VM, but also
+ *    includes customer_metadata changes made through the mdata-put and
+ *    mdata-delete operations from within.
+ *
+ *  * When VMs themselves experience the unexpected. Whether it is a KVM VM has
+ *    its qemu process crash sending the zone to the 'stopped' state or an
+ *    LX/docker container has its init crash without a restart policy in place,
+ *    there are many ways a container can change state without there being a
+ *    corresponding API action.
+ *
+ * vm-agent must notice changes to the VMs on the CN and report these to VMAPI.
+ *
+ *
+ * Theory of Operation
+ * ===================
+ *
+ * On startup vm-agent will:
+ *
+ *  * Start a watcher which will notify vm-agent of VM create/delete/modify
+ *
+ *  * Load the set of VMs that VMAPI thinks should be on this CN
+ *
+ *  * Load the set of VMs that are on this machine (vmadm lookup)
+ *
+ *  * Compare the two lists and:
+ *
+ *     * If there are VMs that exist on the CN, but not VMAPI or have different
+ *       properties between the two: the VM object will be updated in VMAPI.
+ *
+ *     * If there are VMs that exist in VMAPI but not on the CN, the object
+ *       returned by VMAPI will be modified to have state=destroyed and
+ *       zone_state=destroyed and PUT back into VMAPI.
+ *
+ * The initial update to VMAPI is currently done through the:
+ *
+ *     PUT /vms?server_uuid=<uuid>
```

Comment:
If I remember correctly, at least [some comments in VMAPI's repository](https://github.com/joyent/sdc-vmapi/blob/master/lib/endpoints/vms.js#L1113-L1115) indicate that the original goal for the `PUT /vms` endpoint was to _replace_ vms. We should probably at least update these comments, and make sure that the same assumption is not used elsewhere.